### PR TITLE
[Feat][Elementwise] Add 30 UnaryKernel template ops

### LINKED ISSUE #437: [FEAT][ELEMENTWISE] UnaryKernel template ops (30 ops)
## 1. Feature Scope

30 UnaryKernel subclasses. Each is 3-5 lines defining only `op_func(x)`. `UnaryKernel` base class from #436 handles strategy/config/forward.

| Category | Ops | Count |
|----------|-----|-------|
| unary_math | exp, log, sqrt, rsqrt, abs, neg, reciprocal, sign, sin, cos, floor, ceil, round, trunc, erf, log1p, expm1 | 17 |
| activation | gelu, silu, sigmoid, tanh, hardswish, hardsigmoid, mish, selu | 8 |
| logical | logical_not | 1 |
| bitwise | bitwise_not | 1 |
| special | isnan, isinf, isfinite | 3 |

**Excluded**: relu (#436), leaky_relu/elu/hardtanh/softplus/prelu (#439).

### Files

| File | Action |
|------|--------|
| `tileops/kernels/elementwise.py` | Add 30 kernel subclasses |
| `tileops/ops/elementwise.py` | Add 30 op subclasses |
| `tests/ops/test_unary_math.py` | New — 17 ops |
| `tests/ops/test_activation.py` | Append 8 ops to existing |
| `tests/ops/test_logical.py` | New — logical_not |
| `tests/ops/test_bitwise.py` | New — bitwise_not |
| `tests/ops/test_special_elementwise.py` | New — isnan, isinf, isfinite |

### Subclass Pattern

```python
class ExpKernel(UnaryKernel):
    @staticmethod
    def op_func(x):
        return T.exp(x)

class ExpOp(UnaryOp):
    _op_name = "exp"
    kernel_cls = ExpKernel
```

No custom `__init__` or `default_config`. All 30 ops follow this exact pattern.

### op_func Reference

| Kernel | `op_func(x)` return | Notes |
|--------|---------------------|-------|
| Exp | `T.exp(x)` | |
| Log | `T.log(x)` | |
| Sqrt | `T.sqrt(x)` | |
| Rsqrt | `T.rsqrt(x)` | |
| Abs | `T.abs(x)` | |
| Neg | `-x` | |
| Reciprocal | `1.0 / x` | |
| Sign | `T.if_then_else(x > 0, 1.0, T.if_then_else(x < 0, -1.0, 0.0))` | No `T.sign` builtin |
| Sin | `T.sin(x)` | |
| Cos | `T.cos(x)` | |
| Floor | `T.floor(x)` | |
| Ceil | `T.ceil(x)` | |
| Round | `T.round(x)` | |
| Trunc | `T.trunc(x)` | Fallback: `T.if_then_else(x >= 0, T.floor(x), T.ceil(x))` |
| Erf | `T.erf(x)` | |
| Log1p | `T.log(x + 1.0)` | No `T.log1p`; precision risk near 0 |
| Expm1 | `T.exp(x) - 1.0` | No `T.expm1`; precision risk near 0 |
| Gelu | `x * 0.5 * (1.0 + T.erf(x * 0.7071067811865476))` | |
| Silu | `x * T.sigmoid(x)` | Fallback: `x / (1.0 + T.exp(-x))` |
| Sigmoid | `T.sigmoid(x)` | Fallback: `1.0 / (1.0 + T.exp(-x))` |
| Tanh | `T.tanh(x)` | |
| Hardswish | `x * T.min(T.max(x + 3.0, 0.0), 6.0) / 6.0` | |
| Hardsigmoid | `T.min(T.max(x / 6.0 + 0.5, 0.0), 1.0)` | Fixed constants |
| Mish | `x * T.tanh(T.log(1.0 + T.exp(x)))` | |
| Selu | `scale * T.if_then_else(x > 0, x, alpha * (T.exp(x) - 1.0))` | alpha=1.6732632423543772, scale=1.0507009873554805 |
| LogicalNot | `x == 0` | Bool output |
| BitwiseNot | `~x` | Integer input |
| Isnan | `x != x` | Bool output |
| Isinf | `T.isinf(x)` | Fallback: `T.abs(x) == T.infinity()` |
| Isfinite | `(x == x) & (~T.isinf(x))` | Bool output |

If a builtin exists, prefer it over the composite fallback.

### L1 — Correctness (30 cases)

All 30 ops, fp16, 1M elements, same-shape, default strategy. Reference: PyTorch equivalent in fp32 then cast.

| Test File | Ops | PyTorch Reference |
|-----------|-----|-------------------|
| test_unary_math | exp...expm1 (17) | `torch.exp`, `torch.log`, ..., `torch.expm1` |
| test_activation | gelu...selu (8) | `F.gelu`, `F.silu`, `torch.sigmoid`, `torch.tanh`, `F.hardswish`, `F.hardsigmoid`, `F.mish`, `F.selu` |
| test_logical | logical_not (1) | `torch.logical_not` |
| test_bitwise | bitwise_not (1) | `torch.bitwise_not` |
| test_special | isnan, isinf, isfinite (3) | `torch.isnan`, `torch.isinf`, `torch.isfinite` |

Input domain: `randn` for most; `abs(randn) + eps` for log/sqrt/rsqrt.

### L4 — Edge Cases (10 cases, fp32, 4K)

| Op | Input | Expected |
|----|-------|----------|
| sqrt | `[-1, 0, 1e-38, 1]` | NaN, 0, ~1e-19, 1 |
| rsqrt | `[-1, 0, 1e-38, 1]` | NaN, Inf, ~1e19, 1 |
| log | `[-1, 0, 1e-38, 1]` | NaN, -Inf, ~-87.3, 0 |
| log1p | `[-2, -1, 0, 1e-7]` | NaN, -Inf, 0, ~1e-7 |
| exp | `[0, 88.8, -88.8, 200]` | 1, ~large, ~0, Inf |
| expm1 | `[0, 88.8, -88.8, 1e-7]` | 0, ~large, -1, ~1e-7 |
| erf | `[0, 3, -3, 100]` | 0, ~1, ~-1, 1 |
| reciprocal | `[0, 1, -1, 1e-38]` | Inf, 1, -1, ~1e38 |
| sign | `[-5, 0, 3, NaN]` | -1, 0, 1, 0 (or NaN) |
| isnan/isinf/isfinite | `[1, NaN, Inf, -Inf]` | Match PyTorch bool output |

Tolerance: fp32 atol=1e-5 rtol=1e-5, fp16 atol=1e-3 rtol=1e-3.

## 4. Risks

| Risk | Mitigation |
|------|------------|
| TileLang missing builtins (trunc, sigmoid, isinf) | Use composite fallbacks; prefer builtins if they exist |
| Bool output ops (logical_not, isnan, isinf, isfinite) may need `OUTPUT_DTYPE` on UnaryKernel | Same pattern as BinaryKernel comparison ops; coordinate with #436 if template change needed |
| log1p/expm1 composite expressions lose precision near 0 | L4 edge case tests will catch; switch to builtins if available |

## 5. Acceptance Criteria

- [ ] 30 kernel + 30 op subclasses implemented
- [ ] `__all__` updated in both `elementwise.py`
- [ ] L1: 30 ops pass (fp16, 1M, same-shape)
- [ ] L4: 10 edge cases pass (fp32, 4K)
- [ ] `__init__.py` not modified (concrete ops not re-exported)

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Implement 30 `UnaryKernel` subclasses and 30 corresponding `UnaryOp` subclasses for elementwise operations, then tighten the implementation to match issue #437 and the follow-up review findings.

**Ops**: `exp`, `log`, `sqrt`, `rsqrt`, `abs`, `neg`, `reciprocal`, `sign`, `sin`, `cos`, `floor`, `ceil`, `round`, `trunc`, `erf`, `log1p`, `expm1`, `gelu`, `silu`, `sigmoid`, `tanh`, `hardswish`, `hardsigmoid`, `mish`, `selu`, `logical_not`, `bitwise_not`, `isnan`, `isinf`, `isfinite`

Closes #437

## Dtype Support Matrix

| Op / API | Input dtypes | Output dtype | Reference baseline |
| --- | --- | --- | --- |
| Unary math ops + activations | `fp16`, `bf16`, `fp32` | same as input | matching `torch` / `torch.nn.functional` unary semantics |
| `logical_not` | `bool`, `uint8`, `int8`, `int16`, `int32`, `int64`, `fp16`, `bf16`, `fp32` | `bool` | `torch.logical_not` |
| `bitwise_not` | `bool`, `uint8`, `int8`, `int16`, `int32`, `int64` | same as input | `torch.bitwise_not` |
| `isnan`, `isinf`, `isfinite` | `fp16`, `bf16`, `fp32` | `bool` | `torch.isnan`, `torch.isinf`, `torch.isfinite` |

## Test Plan

- [x] **AC-1**: 30 kernel subclasses and 30 op subclasses exist for the issue #437 scope.
- [x] **AC-2**: Unary elementwise dtype contracts are explicit in code via `SUPPORTED_DTYPES` / `OUTPUT_DTYPE` where needed.
- [x] **AC-3**: `gelu` uses the standard `erf` formulation instead of the tanh approximation.
- [x] **AC-4**: `logical_not`, `isnan`, `isinf`, and `isfinite` produce torch-style `bool` outputs and tests validate them with exact comparison.
- [x] **AC-5**: `bitwise_not` support is aligned to the current torch contract (`bool/uint8/int8/int16/int32/int64`) and reject-path tests exist for float dtypes.
- [x] **AC-6**: L4 edge-case coverage matches issue #437 for `sqrt`, `rsqrt`, `log`, `log1p`, `exp`, `expm1`, `erf`, `reciprocal`, `sign`, and special predicates.
- [x] **AC-7**: Repository-level templates/docs now require future new ops to include a dtype support matrix, acceptance checklist, and benchmark table in the PR body.
- [x] **AC-8**: A representative unary benchmark file exists in `benchmarks/ops/bench_unary_elementwise.py` and runs on a real GPU machine with small, medium, and large shapes.
- [x] **AC-9**: Top-level `__init__.py` re-export files remain unchanged.
- [x] Elementwise correctness suite passes on the real machine GPUs: `127 passed, 2 warnings in 8.04s`.
- [x] Unary benchmark suite passes on the real machine GPUs: `21 passed, 23 warnings in 56.25s`.

## Benchmark

**Configuration**: H200, Driver `575.57.08`, CUDA `12.8` (torch), torch `2.9.1+cu128`

| N_total | dtype | Op | TileOPs (ms) | Baseline (ms) | Ratio |
| --- | --- | --- | --- | --- | --- |
| 262,144 | fp16 | `exp` | 0.00250 | 0.00607 | 2.43x faster |
| 1,048,576 | fp16 | `exp` | 0.00334 | 0.00768 | 2.30x faster |
| 4,000,000 | fp16 | `exp` | 0.00680 | 0.01032 | 1.52x faster |
| 262,144 | bf16 | `exp` | 0.00251 | 0.00628 | 2.50x faster |
| 1,048,576 | bf16 | `exp` | 0.00334 | 0.00709 | 2.12x faster |
| 4,000,000 | bf16 | `exp` | 0.00669 | 0.01091 | 1.63x faster |
| 262,144 | fp16 | `gelu` | 0.00264 | 0.00734 | 2.78x faster |
| 1,048,576 | fp16 | `gelu` | 0.00401 | 0.00839 | 2.09x faster |
| 4,000,000 | fp16 | `gelu` | 0.00980 | 0.01323 | 1.35x faster |
| 262,144 | bf16 | `gelu` | 0.00267 | 0.00656 | 2.46x faster |
| 1,048,576 | bf16 | `gelu` | 0.00409 | 0.00845 | 2.07x faster |
| 4,000,000 | bf16 | `gelu` | 0.01009 | 0.01277 | 1.27x faster |
| 262,144 | fp16 | `logical_not` | 0.00262 | 0.00679 | 2.59x faster |
| 1,048,576 | fp16 | `logical_not` | 0.00545 | 0.00691 | 1.27x faster |
| 4,000,000 | fp16 | `logical_not` | 0.01467 | 0.00941 | 0.64x |
| 262,144 | int32 | `bitwise_not` | 0.00297 | 0.00715 | 2.41x faster |
| 1,048,576 | int32 | `bitwise_not` | 0.00604 | 0.00816 | 1.35x faster |
| 4,000,000 | int32 | `bitwise_not` | 0.01657 | 0.01404 | 0.85x |
| 262,144 | fp16 | `isnan` | 0.00263 | 0.00674 | 2.56x faster |
| 1,048,576 | fp16 | `isnan` | 0.00546 | 0.00690 | 1.26x faster |
| 4,000,000 | fp16 | `isnan` | 0.01473 | 0.01016 | 0.69x |

### CODE REVIEW

File: tileops/kernels/elementwise/unary_ops.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    def op_func(x):
+        return T.cast(1.0, "float32") / x
+
+
+class SignKernel(UnaryKernel):
+    """Element-wise sign(x): -1, 0, or +1."""
+
+    @staticmethod
+    def op_func(x):
+        return T.if_then_else(x > 0, T.cast(1.0, "float16"),
```
Comment (zhen8838) [inline tileops/kernels/elementwise/unary_ops.py]: Can't specific float16 here

**Review by zhen8838 — CHANGES_REQUESTED** [review]

File: tileops/kernels/elementwise/unary_ops.py (hunk 1 above)
Comment (lcy-seso) [reply tileops/kernels/elementwise/unary_ops.py]: Good point. The issue here is not only `float16` specifically, but hard-coding any concrete float type in this templated kernel.

I updated `SignKernel` and the related predicate kernels (`LogicalNotKernel` / `IsnanKernel` / `IsinfKernel` / `IsfiniteKernel`) to cast constants to `x.dtype` instead, so the result dtype stays consistent for `fp16` / `fp32` / `bf16` inputs.

I only kept `float32` in kernels where it is used intentionally as an intermediate for numerical stability or lowering constraints.
